command/agent/consul/client.go

```diff
@@ -84,6 +85,100 @@ type operations struct {
 	deregChecks   []string
 }
 
+// AllocRegistration holds the status of services registered for a particular
+// allocations by task.
+type AllocRegistration struct {
+	// Tasks maps the name of a task to its registered services and checks
+	Tasks map[string]*TaskRegistration
+}
+
+func (a *AllocRegistration) Copy() *AllocRegistration {
+	c := &AllocRegistration{
+		Tasks: make(map[string]*TaskRegistration, len(a.Tasks)),
+	}
+
+	for k, v := range a.Tasks {
+		c.Tasks[k] = v.Copy()
+	}
+
+	return c
+}
+
+// NumServices returns the number of registered services
+func (a *AllocRegistration) NumServices() int {
+	if a == nil {
+		return 0
+	}
+
+	total := 0
+	for _, treg := range a.Tasks {
+		for _, sreg := range treg.Services {
+			if sreg.Service != nil {
+				total++
+			}
+		}
+	}
+
+	return total
+}
+
+// NumChecks returns the number of registered checks
+func (a *AllocRegistration) NumChecks() int {
+	if a == nil {
+		return 0
+	}
+
+	total := 0
+	for _, treg := range a.Tasks {
+		for _, sreg := range treg.Services {
+			total += len(sreg.Checks)
+		}
+	}
+
+	return total
+}
+
+// TaskRegistration holds the status of services registered for a particular
+// task.
+type TaskRegistration struct {
+	Services map[string]*ServiceRegistration
+}
+
+func (t *TaskRegistration) Copy() *TaskRegistration {
+	c := &TaskRegistration{
+		Services: make(map[string]*ServiceRegistration, len(t.Services)),
+	}
+
+	for k, v := range t.Services {
+		c.Services[k] = v.Copy()
+	}
+
+	return c
+}
+
+// ServiceRegistration holds the status of a registered Consul Service and its
+// Checks.
+type ServiceRegistration struct {
+	// serviceID and checkIDs are internal fields that track just the IDs of the
+	// services/checks registered in Consul. It is used to materialize the other
+	// fields when queried.
+	serviceID string
+	checkIDs  map[string]struct{}
+
+	// Service is the AgentService registered in Consul.
+	Service *api.AgentService
+
+	// Checks is the status of the registered checks.
+	Checks []*api.AgentCheck
+}
+
+func (s *ServiceRegistration) Copy() *ServiceRegistration {
+	return &ServiceRegistration{
+		serviceID: s.serviceID,
+		checkIDs:  helper.CopyMapStringStruct(s.checkIDs),
+	}
```

Comment:
If not copying Services and Checks is intentional please add a comment why they're intentionally omitted.